qiskit/transpiler/preset_passmanagers/level0.py

```diff
@@ -107,98 +88,49 @@ def _choose_layout_condition(property_set):
-        ]
-    else:
-        raise TranspilerError("Invalid translation method %s." % translation_method)
-
-    # 6. Fix any bad CX directions
-    _direction_check = [CheckGateDirection(coupling_map)]
-
-    def _direction_condition(property_set):
-        return not property_set["is_direction_mapped"]
-
-    _direction = [GateDirection(coupling_map)]
-
-    # 7. Unify all durations (either SI, or convert to dt if known)
-    # Schedule the circuit only when scheduling_method is supplied
-    _scheduling = [TimeUnitConversion(instruction_durations)]
-    if scheduling_method:
-        if scheduling_method in {"alap", "as_late_as_possible"}:
-            _scheduling += [ALAPSchedule(instruction_durations)]
-        elif scheduling_method in {"asap", "as_soon_as_possible"}:
-            _scheduling += [ASAPSchedule(instruction_durations)]
-        else:
-            raise TranspilerError("Invalid scheduling method %s." % scheduling_method)
-
-    # 8. Call measure alignment. Should come after scheduling.
-    _alignments = [
-        ValidatePulseGates(
-            granularity=timing_constraints.granularity, min_length=timing_constraints.min_length
-        ),
-        AlignMeasures(alignment=timing_constraints.acquire_alignment),
-    ]
-
     # Build pass manager
-    pm0 = PassManager()
     if coupling_map or initial_layout:
-        pm0.append(_given_layout)
-        pm0.append(_choose_layout, condition=_choose_layout_condition)
-        pm0.append(_embed)
-        pm0.append(_unroll3q)
-        pm0.append(_swap_check)
-        pm0.append(_swap, condition=_swap_condition)
-    pm0.append(_unroll)
+        layout = PassManager()
+        layout.append(_given_layout)
+        layout.append(_choose_layout, condition=_choose_layout_condition)
+        layout += common.generate_embed_passmanager(coupling_map)
+        routing = common.generate_routing_passmanager(routing_pass, coupling_map)
+    else:
+        layout = None
+        routing = None
+    translation = common.generate_translation_passmanager(
+        basis_gates, translation_method, approximation_degree
+    )
     if coupling_map and not coupling_map.is_symmetric:
-        pm0.append(_direction_check)
-        pm0.append(_direction, condition=_direction_condition)
-        pm0.append(_unroll)
-    if inst_map and inst_map.has_custom_gate():
-        pm0.append(PulseGates(inst_map=inst_map))
-    pm0.append(_scheduling)
-    pm0.append(_alignments)
-    return pm0
+        pre_opt = common.generate_pre_op_passmanager(coupling_map)
+        pre_opt += translation
+    else:
+        pre_opt = None
+    sched = common.generate_scheduling_post_opt(
+        instruction_durations, scheduling_method, timing_constraints, inst_map
+    )
+    return FullPassManager(
+        layout=layout,
+        routing=routing,
+        translation=translation,
+        pre_optimization=pre_opt,
+        scheduling=sched,
+    )
```

Comment:
I think I did it this way because the control flow and logic around layout was different between the optimization levels. Like 0 is:

```
if initial_layout is not None:
    layout = initial_layout
else:
    layout = tivial_layout()
```

but level 1 is:

```
if initial_layout is not None:
    layout = initial_layout
elif trivial_layout().is_perfect():
    layout = trivial_layout()
else:
    layout = DenseLayout()
```

but 2 and 3 are:

```
if initial_layout is not None:
    layout = initial_layout
elif trivial_layout().is_perfect():
    layout = trivial_layout()
elif csplayout.converges():
   layout = csplayout()
else:
    layout = DenseLayout()
```

(where the parameters for csp differ)

I was trying to keep things in common isolated to just the pieces which are shared completely between each optimization level. That way you can got to a specific optimization level's code and see what is shared and what is unique.